xde/xde-link/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+            flags: TYPE_EXACT | DRV_EXACT,
+            interpose_lvl: ILEVEL_0,
+            callback_fcn: create_xde_link,
+        },
+    };
+
+/// devfsadm plugin link removal registration
+#[no_mangle]
+pub static _devfsadm_remove_reg: _devfsadm_remove_reg_t =
+    _devfsadm_remove_reg_t {
+        version: DEVFSADM_V0,
+        count: 1,
+        tblp: &devfsadm_remove_t {
+            device_class: b"pseudo\0".as_ptr().cast(),
+            dev_dirs_re: b"^xde$\0".as_ptr().cast(),
+            flags: RM_HOT | RM_PRE | RM_ALWAYS,
+            interpose_lvl: ILEVEL_0,
+            callback_fcn: devfsadm_rm_all,
+        },
+    };
+
+/// Create xde /dev link for the control device
+///     /dev/xde => /devices/pseudo/xde@0:ctl
+unsafe extern "C" fn create_xde_link(
+    minor: *const di_minor,
+    node: *const di_node,
+) -> c_int {
+    #[link(name = "c")]
+    extern "C" {
+        fn strcmp(s1: *const c_char, s2: *const c_char) -> c_int;
+    }
+    if strcmp(di_minor_name(minor), b"ctl\0".as_ptr().cast()) == 0 {
+        devfsadm_mklink(b"xde\0".as_ptr().cast(), node, minor, 0);
+    }
+    0
+}
+
+use core::ffi::c_char;
+use core::ffi::c_int;
+use core::ffi::c_uint;
+
+/// devfsadm plugin interface version 0
+pub const DEVFSADM_V0: c_uint = 0;
+
+/// Create /dev link at the root
+pub const ILEVEL_0: c_int = 0;
+
+/// Match minor node type exactly
+pub const TYPE_EXACT: c_int = 0x01;
+
+/// Match driver name exactly
+pub const DRV_EXACT: c_int = 0x10;
+
+/// Remove /dev link when device is hot-removed
+pub const RM_HOT: c_int = 0x01;
+
+/// Remove /dev link before processing entire devinfo tree
+pub const RM_PRE: c_int = 0x02;
+
+/// Remove /dev link even if cleanup wasn't requested
+pub const RM_ALWAYS: c_int = 0x08;
+
+/// Minor node type for pseudo devices
+pub const DDI_PSEUDO: *const c_char = b"ddi_pseudo\0".as_ptr().cast();
+
+/// Opaque minor node handle
+type di_minor = core::ffi::c_void;
+
+/// Opaque device node handle
+type di_node = core::ffi::c_void;
+
+// See lib/libdevinfo/libdevinfo.h
+#[link(name = "devinfo")]
+extern "C" {
+    /// Returns name for give minor node
+    fn di_minor_name(minor: *const di_minor) -> *const c_char;
+}
+
+// These symbols exist in the `devfsadm` binary itself which is the one
+// that will be `dlopen()`'ing the plugin.
+extern "C" {
+    fn devfsadm_mklink(
+        link: *const c_char,
+        node: *const di_node,
+        minor: *const di_minor,
+        flags: c_int,
+    ) -> c_int;
+    fn devfsadm_rm_all(file: *const c_char);
+}
+
+/// Predicates used to match a device and how to create its /dev link
+///
+/// cmd/devfsadm/devfsadm.h
+#[repr(C)]
+struct devfsadm_create_t {
+    /// Device class to match (e.g. "pseudo", "disk")
+    device_class: *const c_char,
+    /// Minor node type to match (e.g. DDI_PSEUDO, DDI_NT_BLOCK)
+    node_type: *const c_char,
+    /// Driver name to match
+    drv_name: *const c_char,
+    /// Flags to control matching
```

Comment:
Same const question as above, though I may be misreading what the const applies to.

Reply:
They effectively are const on the C side too. Every devfsadm plugin uses string literals to initialize these values and I can't imagine anything good happening if someone tried to modify them.

On the rust side, codegen-wise `*mut` and `*const` are treated the same and you can freely cast between them. (There's a difference in variance which isn't relevant here).

Reply:
I guess given what it is in practice, that kind of makes sense. I guess it's just surprising a bit to see us assume that when it's not in the code. I know a lot of stuff isn't great about assuming that, but it does make me thing therefore that it should be and we should file an illumos issue to make these things const and promise what we're actually relying on in practice.